core/arch/arm/mm/core_mmu_lpae.c

```diff
@@ -191,19 +191,56 @@
 #else
 #define NUM_L1_TABLES	1
 #endif
-uint64_t l1_xlation_table[NUM_L1_TABLES][CFG_TEE_CORE_NB_CORE][NUM_L1_ENTRIES]
+
+typedef uint64_t l1_xlat_tbls_t[NUM_L1_ENTRIES][CFG_TEE_CORE_NB_CORE];
+typedef uint64_t xlat_tbl_t[XLAT_TABLE_ENTRIES];
+
+l1_xlat_tbls_t l1_xlation_table[NUM_L1_TABLES]
 	__aligned(NUM_L1_ENTRIES * XLAT_ENTRY_SIZE) __section(".nozi.mmu.l1");
 
-static uint64_t xlat_tables[MAX_XLAT_TABLES][XLAT_TABLE_ENTRIES]
+static xlat_tbl_t xlat_tables[MAX_XLAT_TABLES]
 	__aligned(XLAT_TABLE_SIZE) __section(".nozi.mmu.l2");
 
+#define XLAT_TABLES_SIZE	(sizeof(xlat_tbl_t) * MAX_XLAT_TABLES)
+
 /* MMU L2 table for TAs, one for each thread */
-static uint64_t xlat_tables_ul1[CFG_NUM_THREADS][XLAT_TABLE_ENTRIES]
+static xlat_tbl_t xlat_tables_ul1[CFG_NUM_THREADS]
 	__aligned(XLAT_TABLE_SIZE) __section(".nozi.mmu.l2");
 
-static unsigned int next_xlat __kbss;
 static paddr_t max_pa __kbss;
-static int user_va_idx __kdata = -1;
+
+struct mmu_partition {
+	l1_xlat_tbls_t *l1_tables;
+	xlat_tbl_t *xlat_tables;
+	xlat_tbl_t *l2_ta_tables;
+	unsigned int xlat_tables_used;
+	unsigned int total_xlat_tables;
+	int user_va_idx;
+	unsigned int asid;
+};
+
+static struct mmu_partition default_partition __kdata = {
+	.l1_tables = l1_xlation_table,
+	.xlat_tables = xlat_tables,
+	.l2_ta_tables = xlat_tables_ul1,
+	.xlat_tables_used = 0,
+	.total_xlat_tables = MAX_XLAT_TABLES,
+	.user_va_idx = -1,
+	.asid = 0
+};
+
+#ifdef CFG_VIRTUALIZATION
+static struct mmu_partition *current_prtn[CFG_TEE_CORE_NB_CORE] __kbss;
+#endif
+
+static struct mmu_partition *get_prtn(void)
+{
+#ifdef CFG_VIRTUALIZATION
+	return current_prtn[get_core_pos()];
```

Comment:
It looks like this function only is called with exceptions masked so it will work.
I think code could be simplified if this function did mask exceptions itself since the returned value of this function will be constant for a particular thread (or even the active partition).